### DISCUSSION TITLE
STM32L0 configured

### DISCUSSION THREAD

**Discussion by MikeAtETI** [jefftenney/LPTIM-Tick#9] [Q&A]
Hi,

Has anyone configured this for an STM32L0 - I thought I'd stepped through the changes to vUlpPreSleepProcessing and vUlpPostSleepProcessing but I can't get it working reliably:

```
void vUlpPreSleepProcessing(TickType_t* value)
{
   int useDeepSleep = pdFALSE;
   if (xDeepSleepForbiddenFlags == 0)
   {
      useDeepSleep = pdTRUE;
     // MODIFY_REG(PWR->CR, PWR_CR1_LPMS_Msk, PWR_CR1_LPMS_STOP2);
   }
   else if ((xDeepSleepForbiddenFlags & ~ulpPERIPHERALS_OK_IN_STOP1) == 0)
   {
      useDeepSleep = pdTRUE;
    //  MODIFY_REG(PWR->CR, PWR_CR1_LPMS_Msk, PWR_CR1_LPMS_STOP1);
   }

   if (useDeepSleep)
   {
	      uint32_t tmpreg = 0U;
	      uint32_t ulpbit, vrefinbit;
      rccCrSave = RCC->CR;
      rccCfgrSave = RCC->CFGR;

      HAL_PWREx_EnableUltraLowPower();
      HAL_PWREx_EnableFastWakeUp();

      /* It is forbidden to configure both EN_VREFINT=1 and ULP=1 if the device is
         in Stop mode or in Sleep/Low-power sleep mode */
      ulpbit = READ_BIT(PWR->CR, PWR_CR_ULP);
      vrefinbit = READ_BIT(SYSCFG->CFGR3, SYSCFG_CFGR3_EN_VREFINT);
      if((ulpbit != 0) && (vrefinbit != 0))
      {
        CLEAR_BIT(PWR->CR, PWR_CR_ULP);
      }

      /* Select the regulator state in Stop mode ---------------------------------*/
      tmpreg = PWR->CR;

      /* Clear PDDS and LPDS bits */
      CLEAR_BIT(tmpreg, (PWR_CR_PDDS | PWR_CR_LPSDSR));

     /* Set LPSDSR bit according to PWR_Regulator value */
      SET_BIT(tmpreg, PWR_LOWPOWERREGULATOR_ON);

      /* Store the new value */
      PWR->CR = tmpreg;

      /* Set SLEEPDEEP bit of Cortex System Control Register */
      SET_BIT(SCB->SCR, SCB_SCR_SLEEPDEEP_Msk);

      /* Select Stop mode entry --------------------------------------------------*/
        /* Request Wait For Interrupt */
        __WFI();

      /* Reset SLEEPDEEP bit of Cortex System Control Register */

        if((ulpbit != 0) && (vrefinbit != 0))
      {
        SET_BIT(PWR->CR, PWR_CR_ULP);
      }

      *value = 0;
      return;
      //SCB->SCR |= SCB_SCR_SLEEPDEEP_Msk;
   }
   *value = 1;
}
```

```
void vUlpPostSleepProcessing(TickType_t value)
{
   if (SCB->SCR & SCB_SCR_SLEEPDEEP_Msk)
   {
      #ifdef configMIN_RUN_BETWEEN_DEEP_SLEEPS    // Errata workaround
      {
         //     Start a new SysTick timer period.  We won't attempt to enter STOP mode until the timer period
         // ends.  Note that we do start a new period here unnecessarily if the CPU didn't actually enter stop
         // mode (due to a pending interrupt).  That's OK.
         //
         SysTick->VAL = 0;
         SysTick->CTRL = SysTick_CTRL_CLKSOURCE_Msk | SysTick_CTRL_ENABLE_Msk | SysTick_CTRL_TICKINT_Msk;

         //      Mark the "min-run" peripheral as being now in use so that we won't attempt STOP mode until
         // it's no longer in use.  See SysTick_Handler() below.
         //
         xDeepSleepForbiddenFlags |= ulpPERIPHERAL_MIN_RUN;
      }
      #endif

      //      We may have been in deep sleep.  If we were, the hardware cleared several enable bits in the CR,
      // and it changed the selected system clock in CFGR.  Restore them now.  If we're restarting the PLL as
      // the CPU clock here, the CPU will not wait for it.  Instead, the CPU continues executing from the
      // wake-up clock (MSI in our case) until the PLL is stable and then the CPU starts using the PLL.
      //
      RCC->CR = rccCrSave;
      RCC->CFGR = rccCfgrSave;

      SCB->SCR &= ~SCB_SCR_SLEEPDEEP_Msk;

      //      This application bypasses the RTC shadow registers, so we don't need to clear the sync flag for
      // those registers.  They are always out of sync when coming out of deep sleep.
      //
      // RTC->ISR &= ~RTC_ISR_RSF;
   }
}
```

and 

```
void vPortSetupTimerInterrupt( void )
{
   //      Enable the APB clock to the LPTIM.  Then select either LSE or LSI as the kernel clock for the
   // LPTIM.  Then be sure the LPTIM "freezes" when the debugger stops program execution.  Then reset the
   // LPTIM just in case it was already in use prior to this function.
   //
   //      Modify these statements as needed for your STM32.  See LPTIM Instance Selection (above) for
   // additional information.
   //
#ifdef STM32L0
   RCC->APB1ENR |= RCC_APB1ENR_LPTIM1EN;
   MODIFY_REG(RCC->CCIPR, RCC_CCIPR_LPTIM1SEL, LPTIMSEL_Val << RCC_CCIPR_LPTIM1SEL_Pos);
   DBGMCU->APB1FZ |= DBGMCU_APB1_FZ_DBG_LPTIMER_STOP;
   RCC->APB1RSTR |= RCC_APB1RSTR_LPTIM1RST;   // Reset the LPTIM module per erratum 2.14.1.
   RCC->APB1RSTR &= ~RCC_APB1RSTR_LPTIM1RST;
#else
   RCC->APB1ENR1 |= RCC_APB1ENR1_LPTIM1EN;
   MODIFY_REG(RCC->CCIPR, RCC_CCIPR_LPTIM1SEL, LPTIMSEL_Val << RCC_CCIPR_LPTIM1SEL_Pos);
   DBGMCU->APB1FZR1 |= DBGMCU_APB1FZR1_DBG_LPTIM1_STOP;
   RCC->APB1RSTR1 |= RCC_APB1RSTR1_LPTIM1RST;   // Reset the LPTIM module per erratum 2.14.1.
   RCC->APB1RSTR1 &= ~RCC_APB1RSTR1_LPTIM1RST;
#endif
   #ifdef STM32WL   // <-- "Family" symbol is defined in the ST device header file, e.g., "stm32wlxx.h".
   {
      #define EXTI_IMR1_LPTIM1   (1UL << 29)
      #define EXTI_IMR1_LPTIM2   (1UL << 30)
      #define EXTI_IMR1_LPTIM3   (1UL << 31)

      //      Users of STM32WL must also change this next statement to match their LPTIM instance selection.
      // By default these MCU's disable wake-up from deep sleep via LPTIM.  An oversight by ST?
      //
      EXTI->IMR1 |= EXTI_IMR1_LPTIM1;
   }
   #endif

   //      Be sure the reference clock is ready.  If this assertion fails, be sure your application code
   // starts the reference clock (LSE or LSI) prior to starting FreeRTOS.
   //
   configASSERT(IS_REF_CLOCK_READY());

   //      Calculate the constants required to configure the tick interrupt.
   //
   ulTimerCountsForOneTick = ( LPTIM_CLOCK_HZ + ( configTICK_RATE_HZ / 2 ) ) / configTICK_RATE_HZ;
   configASSERT( ulTimerCountsForOneTick >= 4UL );  // CLOCK frequency must be at least 3.5x TICK frequency

   //      Calculate the maximum number of ticks we can suppress.  Give 1 OS tick of margin between clearly
   // future match events and clearly past match events.  Anything within the previous one tick is clearly
   // past, within one tick before that is in the margin between, which we call the past for convenience.
   // Everything else is in the future when isTickNowSuppressed is true and in the past otherwise.  And
   // set up a couple of other things for the precision feature, if enabled.

   #if ( configLPTIM_ENABLE_PRECISION == 0 )
   {
      xMaximumSuppressedTicks = 65536UL / ulTimerCountsForOneTick - 1 - 1;
   }
   #else
   {
      xMaximumSuppressedTicks = 65536UL * configTICK_RATE_HZ / LPTIM_CLOCK_HZ - 1 - 1;

      //      For convenience, the code above rounded *up* if the ideal number of counts per tick is exactly
      // X.5.  So we might calculate lSubcountErrorPerTick to be -(configTICK_RATE_HZ/2) but never
      // +(configTICK_RATE_HZ/2).  If you get a build error on this line, be sure configTICK_RATE_HZ is a
      // simple numeric literal (eg, 1000UL) with no C typecasting (eg, (TickType_t)1000).
      //
      #if ( LPTIM_CLOCK_HZ % configTICK_RATE_HZ < configTICK_RATE_HZ/2 )
         #define IS_SUBCOUNT_EPT_POSITIVE 1
      #else
         #define IS_SUBCOUNT_EPT_POSITIVE 0
      #endif

      lSubcountErrorPerTick = LPTIM_CLOCK_HZ - ( ulTimerCountsForOneTick * configTICK_RATE_HZ );
      configASSERT( lSubcountErrorPerTick != configTICK_RATE_HZ / 2 );
   }
   #endif // configLPTIM_ENABLE_PRECISION


   //      Configure and start LPTIM.
   //
   LPTIM->IER = LPTIM_IER_CMPMIE | LPTIM_IER_CMPOKIE;   // Modify this register only when LPTIM is disabled.
   LPTIM->CFGR = (0 << LPTIM_CFGR_PRESC_Pos);           // Modify this register only when LPTIM is disabled.
   LPTIM->CR = LPTIM_CR_ENABLE;
   LPTIM->ARR = 0xFFFF;        // timer period = ARR + 1.  Modify this register only when LPTIM is enabled.
   LPTIM->CMP = ulTimerCountsForOneTick;                // Modify this register only when LPTIM is enabled.
   isCmpWriteInProgress = pdTRUE;
   usIdealCmp = ulTimerCountsForOneTick;
   #if ( configLPTIM_ENABLE_PRECISION != 0 )
   {
      lRunningSubcountError = lSubcountErrorPerTick;
   }
   #endif // configLPTIM_ENABLE_PRECISION
   LPTIM->CR |= LPTIM_CR_CNTSTRT;

   //      Enable the timer interrupt at the configured priority.  See configTICK_INTERRUPT_PRIORITY for
   // important details.
   //
   NVIC_SetPriority( LPTIM_IRQn, configTICK_INTERRUPT_PRIORITY );
   NVIC_EnableIRQ( LPTIM_IRQn );
}
```

**Comment by Scarittagle** [comment]
I know this post is old but if you are using MXCube to generate code base, STM32L0 uses FreeRTOS Version 10.2.1 instead of other MCUs use newer 10.3.1 (as of July 2024). And in 10.2.1, the Setup Timer Interrupt function is defined as `void prvSetupTimerInterrupt( void )` instead of `__attribute__(( weak )) void vPortSetupTimerInterrupt( void )`, that's why it wouldn't work: it is not calling the function at all.

You will need to modify it in port.c in order to get it to work.

**Reply by jefftenney** [reply]
Thanks @Scarittagle, that is useful information for users of CubeMX.  On STM32L0, CubeMX generates FreeRTOS kernel files too old for lptimTick.c.

For perspective, *all* users of lptimTick.c should be using FreeRTOS kernel v10.4 or newer, as noted in [this comment](https://github.com/jefftenney/LPTIM-Tick/blob/6dc0a7d6ebcb2d449efff60f0ba85bca24fdb87c/Core/Src/lptimTick.c#L368-L369).

If you are developing a project today using STM32L0 and CubeMX, I would suggest integrating the newest version of FreeRTOS "manually".  That means you would *not* tell CubeMX that you are using FreeRTOS.  Just add the kernel and port files manually into your project, and then create your tasks and start the OS from a "User Code" section in main().  See [LPTIM-Tick-U5](https://github.com/jefftenney/LPTIM-Tick-U5) as an example.  (Heads up - that example uses the port files for Cortex M33 not Cortex M0.)

**Reply by Scarittagle** [reply]
Yes, I was trying to integrate it with 10.2.1 and it does not work quite well even with the modified function. It will throw me into the hard fault when ever osDelay() related function is used.

**Reply by SaiVenkatB** [reply]
Hey @jefftenney I am trying to integrate the code for stm32L0 for some reason code is getting struck at osdelay. I have even updated Freertos to newest version still  I am not able to make it work. However I have seen a person ported his code to work on WLE, I used for my stm32wl55 it is working well. But with STM32L0 I am having no luck. 

Things I have changed are 
FreeRTOSConfig.h same as LPTIM-Tick

lptimTick.c
```
RCC->APB1ENR |= RCC_APB1ENR_LPTIM1EN;
MODIFY_REG(RCC->CCIPR, RCC_CCIPR_LPTIM1SEL, LPTIMSEL_Val << RCC_CCIPR_LPTIM1SEL_Pos);
DBGMCU->APB1FZ |= DBGMCU_APB1_FZ_DBG_LPTIMER_STOP;
RCC->APB1RSTR |= RCC_APB1RSTR_LPTIM1RST;   // Reset the LPTIM module per erratum 2.14.1.
RCC->APB1RSTR &= ~RCC_APB1RSTR_LPTIM1RST;
```
ulp.c 

```
void vUlpPreSleepProcessing()
{
   int useDeepSleep = pdFALSE;
   if (xDeepSleepForbiddenFlags == 0)
   {
      useDeepSleep = pdTRUE;
      MODIFY_REG(PWR->CR, PWR_CR_LPSDSR_Msk, 1);
   }
   else if ((xDeepSleepForbiddenFlags & ~ulpPERIPHERALS_OK_IN_STOP1) == 0)
   {
      useDeepSleep = pdTRUE;
      MODIFY_REG(PWR->CR, PWR_CR_LPSDSR_Msk, 0);
   }

   if (useDeepSleep)
   {
      rccCrSave = RCC->CR;
      rccCfgrSave = RCC->CFGR;

      SCB->SCR |= SCB_SCR_SLEEPDEEP_Msk;
   }
}
```
I am using an external LSE and TIM7 as time base. 

main.c 

```
void StartDefaultTask(void *argument)
{
  /* USER CODE BEGIN 5 */
	   //      Stop the timer that provides the HAL tick.  Now it won't needlessly interrupt tickless idle periods
	   // that use sleep mode.  (The HAL tick already can't interrupt tickless idle periods that use stop mode,
	   // because the HAL timer doesn't operate in stop mode.)  In a real application, the HAL tick might be
	   // required by the HAL even after FreeRTOS has control.  It's still best to stop the timer here, and then
	   // define HAL_GetTick() and HAL_Delay() to use the FreeRTOS tick (and delay) once available.
	   //
   TIM7->CR1 &= ~TIM_CR1_CEN;  // wish CubeMX would generate a symbol for the HAL tick timer

	  /* Infinite loop */
	  for(;;)
	  {
	    osDelay(5000);
	    for(int j = 0; j < 100 ; j++){
	        for(int i = 0; i < 65535 ; i++){
	        	asm("nop");
	        }
	    }
	  }
  /* USER CODE END 5 */
}
```
Can you give me some suggestions to debug this.

**Reply by jefftenney** [reply]
Sure I can try to help.  

- Is your project using ST's code generator?  (Both CubeIDE and CubeMX can generate code for you.)
- Have you tried without tickless idle (and thus without lptimTick.c)?  You would set configUSE_TICKLESS_IDLE to 0.
- Can you post a minimal project that demonstrates the issue you are facing?

**Reply by b-saivenkat** [reply]
Hello Jeff,

Thank you for your guidance. I was able to identify and resolve the issue.

I am using a FreeRTOS project generated by STM32CubeMX, which currently provides FreeRTOS v10.2. While updating the FreeRTOS sources, I realized that I had overlooked updating the directory `Middlewares/Third_Party/FreeRTOS/Source/portable/GCC/ARM_CM0`. As a result, the `lptimTick.c` file I was using contained a `vPortSetupTimerInterrupt()` implementation that did not match the STM32L0 port, where the corresponding function in `port.c` is named `prvSetupTimerInterrupt()`.

After aligning these and updating the files correctly, STOP mode began working as expected, even with FreeRTOS v10.2.

For reference, I have published the STM32L0 project here:
[https://github.com/b-saivenkat/LPTIM_Tick_STM32l0xx/](https://github.com/b-saivenkat/LPTIM_Tick_STM32l0xx/)

I am currently testing on a custom board and observing a STOP-mode current consumption of approximately 0.3 mA with FreeRTOS. In a bare-metal implementation (without RTOS), I am able to reach around 0.08 mA in STOP mode. I am now investigating the source of this additional current draw in the RTOS-based setup. If you have any insights into common causes for this difference, I would appreciate your thoughts.

Additionally, I wanted to ask about clock configuration. Is it feasible to use HSE instead of MSI for this setup? When I attempted to switch to HSE, a `configASSERT` was triggered, which suggests there may be a known limitation or configuration requirement related to the clock source.

Thank you again for the excellent library and for your continued support.

Best regards,
Sai Venkat

**Reply by jefftenney** [reply]
**High Quiescent Current**

I looked at your repo. In your case, I would probably consider the UART and port-pin settings.  You could temporarily drop the UART completely and see what difference it makes.  Sometimes UARTs have the ability to override the STOP mode and prevent it from stopping the UART's reference clock.  Also I would go through the port-pin settings in your bare-metal solution (0.08 mA) and compare them to the  FreeRTOS solution.  Could be something simple there.

**HSE**

In an ultra low power application, it's best not to use HSE as your permanent system clock.  If there are higher power activities that your device must do with an accurate frequency reference, those could use HSE as the system clock temporarily.  Then when you're not doing the higher power activities, you would still be ultra low power.  It's hard to give specific advice without knowing the final application.  But yes, HSE can be use in ULP applications, just very carefully.

In general, your application should keep track of when it is "ready" for the STOP modes. See ulp.c as an example. Then temporary HSE usage can fit into that pattern.  See guidance here (applicable to the STM32L4 but still useful as a reference for you).  https://github.com/jefftenney/LPTIM-Tick/blob/a63ca68062b81d0f05b83e56297e0d67a94f14e8/Core/Src/ulp.c#L16

**Reply by b-saivenkat** [reply]
Sorry for the late reply — I was testing the system thoroughly.

The system’s current consumption has now come down to 0.08 mA, as observed in bare-metal mode. The issue appears to have resolved without any software changes. I suspect that something may have affected the hardware during the earlier test, which resulted in the higher quiescent current.

On the same hardware, I am now also seeing 0.08 mA with the LPTIM tick enabled. I will try to replicate this behavior on other hardware as well.

I haven’t done extensive testing yet, but when ulpInit() was removed, the system was running on HSE. The system has been running well on MSI without any issues so far; however, in terms of current consumption, I haven’t observed any significant reduction when comparing HSE and MSI.

Thank you for your continued support.
